### PR TITLE
Build Testing approach using tcell simulation screen

### PR DESCRIPTION
**Pull request by lawrencegripper**
**Related issue**
#86

**Describe**

This is WIP looking at building a simple to use integration testing approach for gocui. The aim is to allow input and assertion about the
resulting screen content.

**Todo**

- [x] Update docs to show how to use this to test gocui
- ~~[ ] Add some sample tests to the _examples~~ not got time to complete this item atm.

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# See here for image contents: https://github.com/microsoft/vscode-dev-containers/tree/v0.163.1/containers/go/.devcontainer/base.Dockerfile
```
Comment (lawrencegripper) [inline .devcontainer/Dockerfile]: Ignore this file, will remove - it relates to VSCode dev containers which I use locally.

File: testing.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
+// Copyright 2020 The gocui Authors. All rights reserved.
```
Comment (mjarkk) [inline testing.go]: Copyright 2021 :)

**Comment by mjarkk** [comment]
Oh i'm just thinking of something wouldn't it be nice if we have tests for the examples?  
If we have those users know how to write tests for their gocui programs and we can make our cd/ci stronger.  

And can you add testing capabilities to the readme features section?

**Comment by lawrencegripper** [comment]
Yeah sounds good to me, I'll add a set of tests to one of the examples then reference to this from the docs showing how to run tests. 

Playing around with how/what to expose, tests still look a bit busy so might add a few helper methods to make them nicer to write.

**Comment by lawrencegripper** [comment]
So I'm pretty happy with the way the tests have worked out (except I have a file called `testing_test.go` and some of the naming).

Be great to get some thoughts on this before I write up the docs. The `testing_test.go` should show how this can be used.

**Comment by lawrencegripper** [comment]
Any thoughts on this one? If we're happy with the approach I can finish off the docs and get this ready for review.

File: go.mod [hunk 3]
```diff
@@ -5,4 +5,5 @@ go 1.13
 require (
 	github.com/gdamore/tcell/v2 v2.0.0
 	github.com/mattn/go-runewidth v0.0.9
+	github.com/onsi/gomega v1.11.0
```
Comment (mjarkk) [inline go.mod]: I'm not sure if i want this testing package to be shipped to all library users, can we also test it without this package?

**Review by mjarkk — COMMENTED** [review]
Ya this is looking good!

I've personally not written a lot of test so i'm probably not the one to tell whats right or wrong but from what i can see in the code it looks good and it seems we can expend on this code.  

I would like to have this merged as the possibility to easily write tests is always better than no possibility *(or is very hard)* to test.

File: go.mod (hunk 3 above)
Comment (lawrencegripper) [reply go.mod]: Yeah, we should be able to use the in-built testing library. Let me give it a try

**Comment by lawrencegripper** [comment]
Think this should be all set, I've added some basic docs to the README and removed the usage of `gomega`

File: README.md [hunk 4]
```diff
@@ -111,6 +111,89 @@ func quit(g *gocui.Gui, v *gocui.View) error {
 }
 ```
 
+### Example with Simulated Test Screen
+
+You can write simple tests for `gocui` which let you simulate keyboard and then validate the output drawn to the screen.
+
+1. Create an instance of `gui` with `OutputSimulator` set as the mode `g, err := NewGui(OutputSimulator, true)`
+1. Call `GetTestingScreen` to get a `testingScreen` instance. 
+1. On this you can use `SendKey` to simulate input and `GetViewContent` to evaluate what is drawn.
```
Comment (mjarkk) [inline README.md]: Why is this all 1. ?

File: README.md [hunk 5]
```diff
@@ -111,6 +111,89 @@ func quit(g *gocui.Gui, v *gocui.View) error {
 }
 ```
 
+### Example with Simulated Test Screen
```
Comment (mjarkk) [inline README.md]: I find this title a bit too vague as you have to read the complete title to understand it.  
Personally i would like to know by the first few words "oh this is about testing a gocui program".

Can you maybe change this to be something short and clear like: `Testing example`  
Or change this to more be faster understandable like: `Test gocui program using a simulated screen`?

Also can you change this to not be a parent of the example? `###` > `##`

**Review by mjarkk — CHANGES_REQUESTED** [review]
Code looks good, only some documentation i think could be changed

File: README.md (hunk 4 above)
Comment (lawrencegripper) [reply README.md]: Markdown will auto increment numbers for you. I've tended to be lazy and let it do it, that way if I add something in the middle I don't have to go fix them all up. 

Preview here shows this comes out as [1/2/3](https://github.com/awesome-gocui/gocui/blob/c54069f2067031a052baa502cd48de39fb53ce59/README.md)

File: README.md (hunk 5 above)
Comment (lawrencegripper) [reply README.md]: Yeah sounds good, I'll update this now.

File: README.md (hunk 4 above)
Comment (lawrencegripper) [reply README.md]: I'll fix this up now as you are right, it does make it more readable if you are looking at the markdown directly

File: go.mod (hunk 3 above)
Comment (lawrencegripper) [reply go.mod]: This should be all set now, with the code using gomega removed I ran `go mod tidy` which removed it.

File: README.md (hunk 5 above)
Comment (lawrencegripper) [reply README.md]: It's now `## Testing example`

**Review by mjarkk — APPROVED** [review]

**Comment by mjarkk** [comment]
One last thing. Can you update to the latest master, after that I'll merge it :)

**Comment by lawrencegripper** [comment]
No worries, I've rebased this onto the current `master` now.

**Pull request merged**